paddle/fluid/operators/optimizers/lamb_op.h

```diff
@@ -0,0 +1,316 @@
+  }
+
+  inline HOSTDEVICE void operator()(size_t i) const {
+    auto row_idx =
+        math::BinarySearch<int64_t>(rows_, row_count_, i / row_numel_);
+    T g = row_idx >= 0 ? grad_[row_idx * row_numel_ + i % row_numel_] : 0;
+    update(i, g);
+  }
+};
+
+template <typename T>
+struct LambParamUpateFunctor {
+  const T* lr_;
+  const T* param_;
+  const T* param_norm_;
+  const T* trust_ratio_div_;
+  const T* trust_ratio_div_norm_;
+  T* param_out_;
+
+  LambParamUpateFunctor(const T* lr, const T* param, const T* param_norm,
+                        const T* trust_ratio_div, const T* trust_ratio_div_norm,
+                        T* param_out)
+      : lr_(lr),
+        param_(param),
+        param_norm_(param_norm),
+        trust_ratio_div_(trust_ratio_div),
+        trust_ratio_div_norm_(trust_ratio_div_norm),
+        param_out_(param_out) {}
+
+  inline HOSTDEVICE void operator()(size_t i) const {
+    T lr = *lr_;
+    T p_norm = *param_norm_;
+    T tr_div_norm = *trust_ratio_div_norm_;
+
+    lr *= p_norm / tr_div_norm;
+    param_out_[i] = param_[i] - lr * trust_ratio_div_[i];
+  }
+};
+
+template <typename DeviceContext, typename T>
+class LambOpKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const override {
+    const auto* param_var = ctx.InputVar("Param");
+    PADDLE_ENFORCE(param_var->IsType<framework::LoDTensor>(),
+                   "The Var(%s)'s type should be LoDTensor, "
+                   "but the received is %s",
+                   ctx.Inputs("Param").front(),
+                   framework::ToTypeName(param_var->Type()));
+
+    using paddle::framework::LoDTensor;
+    using paddle::operators::detail::Ref;
+
+    T weight_decay = static_cast<T>(ctx.Attr<float>("weight_decay"));
+    T beta1 = static_cast<T>(ctx.Attr<float>("beta1"));
+    T beta2 = static_cast<T>(ctx.Attr<float>("beta2"));
+    T epsilon = static_cast<T>(ctx.Attr<float>("epsilon"));
+    auto& param = Ref(ctx.Input<LoDTensor>("Param"), "Must set Param.");
+    auto* grad_var = ctx.InputVar("Grad");
+    auto& mom1 = Ref(ctx.Input<LoDTensor>("Moment1"), "Must set Moment1.");
+    auto& mom2 = Ref(ctx.Input<LoDTensor>("Moment2"), "Must set Moment2.");
+    auto& lr =
+        Ref(ctx.Input<LoDTensor>("LearningRate"), "Must set LearningRate.");
+
+    auto& beta1_pow =
+        Ref(ctx.Input<LoDTensor>("Beta1Pow"), "Must set Beta1Pow.");
+    auto& beta2_pow =
+        Ref(ctx.Input<LoDTensor>("Beta2Pow"), "Must set Beta2Pow.");
+
+    auto& param_out =
+        Ref(ctx.Output<LoDTensor>("ParamOut"), "Must set ParamOut.");
+    auto& mom1_out =
+        Ref(ctx.Output<LoDTensor>("Moment1Out"), "Must set Moment1Out.");
+    auto& mom2_out =
+        Ref(ctx.Output<LoDTensor>("Moment2Out"), "Must set Moment1Out.");
+
+    platform::ForRange<DeviceContext> for_range(
+        static_cast<const DeviceContext&>(ctx.device_context()), param.numel());
+    framework::Tensor trust_ratio_div;
```

Comment:
There is a potential bug because the GPU execution is async, this is to say that when the calling of `LambOpKernel::Compuate` returns, the computation of the GPU side may be not finished. But `trust_ratio_div` is placed in stack memory, and the memory is to be freed when the calling of `LambOpKernel::Compuate` returns.

Reply:
I found that there are many files use `Tensor tmp`, but the program doesn't have abnormal results, this is because we use only use one CUDA stream, and all the tasks are placed in this stream.

Reply:
Changed these declarations.